public/app/features/query/state/runRequest.ts

```diff
@@ -221,6 +256,10 @@ export function callQueryMethod(
     return of<DataQueryResponse>({ data: [] });
   }
 
+  if (config.featureToggles.queryMigrations && datasource.migrateQuery) {
+    return migrateQueries(request, datasource, queryFunction);
```

Comment:
Is this something we really want to run every time the query executes?  If the goal is for the editor, should it be an init function in the editor?

We could also put it in the dashboard loader 🤔

Reply:
I didn't have time to comment the PR :) So yes, we need to run migrations in two places:
 - When running a query (the editor doesn't need to be loaded when only viewing a panel).
 - When loading the editor.
Note that this is only for frontend only datasources (backend datasources will run migrations on the backend so only the second point applies).

The dashboard loader / migrator may be a good idea so the migration is only executed once per panel target, the only problem is that migrations may be asynchronous and that probably don't play well with the loading process but I'll take a look.

Reply:
Re: Do migration in the dashboard loader

I have tried to use the `DashboardMigrator` for this but I don't think it's possible. The [updateSchema](https://github.com/grafana/grafana/blob/migrateQueryFrontend/public/app/features/dashboard/state/DashboardMigrator.ts#L97) function is synchronous but for running datasource migrations, I need to retrieve the datasource instance, which is an async call:

```ts
          const ds = await getDataSourceSrv().get(target.datasource);
          if (ds.migrateQuery) {
            ... do migration
```

`updateSchema` cannot be easily made async either because it's used in the [DashboardModel constructor](https://github.com/grafana/grafana/blob/migrateQueryFrontend/public/app/features/dashboard/state/DashboardModel.ts#L167).

Am I missing something? Is there another place I can try to add this migration code?